### ISSUE TITLE
How to proceed with open source specs.

### ISSUE THREAD

**Issue by Djfe** [pakastin/open-source-flash#52]
With the full spec it should be possible to write something to compile games from flash/swf to webassembly.
Or write something in webassembly that can run it (maybe with another intermediary format)

As an Alternative Shumway could be completed so it runs with all content and not only if you're lucky.

What are your suggestions? I'm not familiar with flash so it's likely that there better ways to achieve this :)

**Comment by madovsky** [comment]
yes, it can be a good idea, I think the best way would be a strong converter AS2/AS3 to whatever.
those corporation crooks will change again the whole technology in 10 years again and again, forcing
millions websites and application to reprogram everything just for survival, and they fukcing know that, they want to kill a maximum of websites, it's the part of their agenda...

**Comment by pakastin** [comment]
Good points!

**Comment by madovsky** [comment]
pakastin, thanks to make people awake of what's happening, we are the web, and gave our life to program it, not them.

**Comment by asdfjk1** [comment]
And what about for shockwave, how will we go about it? I really don't want it to die off and I know I'm not the only one.

**Comment by pakastin** [comment]
I started my web career with Flash (Flash 4 in ~1999) and I'm really nostalgic about my oldest projects I did with Flash. I fear we lose access to those.

I remember a lot of cool shockwave experiments and projects back then (not done by me) as well.

I also know that many legendary former Flash developers have talked about open sourcing Flash and making a gallery of old projects.

**Comment by pakastin** [comment]
Let's hope Adobe will give us something. Share the love in Twitter etc..
https://twitter.com/pakastin/status/892042787771879424
😉

**Comment by Djfe** [comment]
obviously we want the Flash and swf (Shockwave) spec, so everything can survive ;) (as said above)

**Comment by Djfe** [comment]
the gallery could be hosted by archive.org the way they already do it with old MS-DOS Software and games (running in HTML5)

**Comment by asdfjk1** [comment]
> obviously we want the Flash and swf (Shockwave) spec, so everything can survive ;) (as said above)

And for the DIR/DCR that shockwave used

**Comment by ajwfrost** [comment]
I'm slightly confused by this request, "we want the Flash .. spec" - but you have this already. Adobe published the spec years ago, and more recently they changed the licensing of the spec so that it's free for people to use even to create an alternative to their Flash Player (e.g. gnash use it). As well as the SWF spec, there's an "AVM2 Overview" which describes the ActionScript virtual machine and the bytecode format. The virtual machine itself is already open source (on github) and there are Flash Player downloads/archives already on Adobe's website.

So what is the request here? If you want to be able to play Flash-based games beyond 2020 then you just need to keep a machine around that doesn't update the browser, there's no "kill switch" in Flash that will stop your previously downloaded version from working. If you want the SWF spec so that people could theoretically create an alternative flash player in the future, then it's there and the avmplus project gives you a great starting point. Even the Flash/AIR APIs are provided in the "playerglobal.swc" and "airglobal.swc" files that come with the freely downloadable AIR SDKs, so I think Adobe already provide everything that anyone would need if they wanted to create an alternative flash player.

Also it's worth checking out their (open source) "SWF Investigator" tool which reads in a SWF file and shows you the tags and all their detail, plus there are a lot of other commercial SWF decompilers.
Finally - there's mention of converters for ActionScript or for SWF files; I put together a converter a while back that took a SWF and converted the ActionScript 3 bytecode from it into C++ that then could be built along with a separate framework into a Windows Phone application. It was very incomplete of course but the concept is sound, so all this is already possible with just the information that's already in the public domain..

I'd be genuinely interested to find out what the goals are here, what do people actually want, to see if it's not something that they could get already. My own thoughts are that, however much we might want to keep the Flash Player plug-in alive, it won't be much good when all of the browsers have removed their support for the format. So I would be more inclined to focus on conversion tools so that SWF content can be converted into something that all the browsers can view "normally"/in the future..


Thanks

**Comment by ROBERT-MCDOWELL** [comment]
agree

**Comment by demurgos** [comment]
Personally, the main reason why I'd like a release of the source code is 
that the SWF spec is sometimes incomplete, inaccurate or even wrong. 
Cross-checking with the Flex source code, Shumway and Gnash can yield 
conflicting results.

The source code is the actual "source of truth" about what the player 
does so being able to read it would solve these documentation issue.

I am planning to craft some SWFs manually to check what are the effects 
of some flags or how the errors are handled, but it will never be as 
rigorous as if I had the source code.

**Comment by Djfe** [comment]
@ajwfrost I didn't even know about avmplus, so thx a lot for your reply! :)

The reason why I thought nothing was available is, because I found nothing. And the development of shumway subsided even though it's incomplete/not working for lots of game swf's I found out there.

I guess, people mostly need to invest more time into optimizing shumway then, right?
(gnash seems nice, but only runs natively, which has disadvantages (security, plattform independency)

Is the avmplus code able to interpret the newest SWF versions? (swf 10+)
Gnash says they don't support SWF 10
@demurgos has a good point here: no specification is better than the actual source code (another good (or rather bad) example is VP9)
though I guess you don't want the source code in public hands, while Flash is still a valid target for attackers.

While it could be a possibility to actually keep an old browser with the official plugin around, it's a hassle, it's not advisable for security reasons and it won't work forever, because who knows if browser x will still run on future OSes and Flash and the browsers are 32bit, which will expire as well, one day.
Virtual Machines are no possibility for DAU's and the community.

> so all this is already possible with just the information that's already in the public domain
the license is MPL-2.0 not public domain, but I guess, you didn't mean that literally.

> to see if it's not something that they could get already
I wonder if there were further reasons why the development of shumway stopped.
Maybe one of the dev's could speak up, if he/she reads this and tell us whether the spec/documentation of flash was too incomplete to continue.
Maybe it was too much work reverse engineering all the details to be feasible.

**Comment by ajwfrost** [comment]
@demurgos yes good point, there were times when they didn't update the SWF spec for a while, and there are likely to be typos etc. But it's a huge beast, I don't know whether access to the source code would help you to find the issues! The AVM2 document is slightly out of date, and doesn't include details of the 'vector' type, so yes for this I needed to look at the SWFInvestigator source to see how these are represented in ActionScript Byte Code. 

@djfe I wasn't actually aware of Shumway, have just taken a look at that. I knew about Swiffy, and of course gnash but yes they didn't pull in the avmplus engine and it didn't really get taken forwards sadly. I still think this is a decision taken as much by the browser vendors as by Adobe, and without a browser that remains capable of loading a plugin then it's going to be of limited use to have a Flash Player plugin; also bear in mind that all companies who currently use Flash on their websites are now going to be transitioning all that to HTML5 over the next couple of years, so I think the requirements to have the Flash Player will go down even further.

By "public domain" I just meant that I could see it :-) rather than looking at the actual licensing terms; MPL is fairly nice I think but the documentation is CC-non-commercial which could be a pain, and the AIR SDK license forbids you decompiling things like airglobal.swc (I think...). Anyway ...

One other consideration is that there is quite a lot of a dependency between the Flash Player and the browser. If there is a "bug" somewhere, then it's actually tricky to determine who's at fault.. and also, you'll find that Adobe sometimes worked around browser bugs in the player (rather than getting MSFT/Mozilla/Opera/etc to fix the browser), which makes things even more complicated. We've seen an error with Flash content from one Firefox tab appearing within another Firefox tab: is it Flash that's at fault here, or Firefox, or GTK..? etc.

I've not looked at Shumway too deeply but I get the impression it's just a Firefox extension rather than something that would work also in Chrome/IE/Edge/Opera/etc? Having something that can interpret/run SWF files directly within a browser would be good but all the current html is sat there with "object" or "embed" tags, so you'd need to change the browsers if you wanted to be able to suddenly handle this within JavaScript/WebAssembly. So to my mind, the much better approach is to create tools that support people in migrating their content to something that the browsers already support. (I saw a note about that on this forum, I'm not sure that Adobe really want/need to spend their own money on this though! so it would need to come from the community or from the companies who have lots of SWF files to migrate into html5....)

Interesting topics :-)   Thanks!

**Comment by Djfe** [comment]
the addon is mostly so that you can run the code on embed/object tags
Shumway itself uses javascript and webgl and stuff like that, so it's browser independent

I guess, whether it will be an interpreter like shumway in the end or a compiler depends on the implementation
both could work
The compiler could have a performance advantage though but it might be more work (and more knowledge is needed)

The issue we all have, aren't companies using flash but the indie scene:
hobby programmers who wrote lots of flash games or made animations, that you can find around the net, which noone (no company) will invest money into to keep them alive.
And the devs aren't around anymore to do it.

The community probably has to do it.

browser bug workarounds are hopefully marked in the source code (or you can see them, if you have the plugin for Firefox and the one for chrome so that you can compare the plugins)

**Comment by ROBERT-MCDOWELL** [comment]
maybe create a browser in AIR with viewport or ANE integrating webkit would be better,
as this we get rid off browser dependency.... ;)

**Comment by pakastin** [comment]
Next question is: how long do you think Air will be updated?

**Comment by ajwfrost** [comment]
:-)   good points above, yes it needs to come from the community and yes it's actually odd that Adobe have been investing so much efforts/money into Flash and AIR that they then give away for free (to most people, although there is still the royalty-bearing licenses for those that don't fit into the EULA terms..)

AIR is a good question. Some of the main advantages I see of AIR over Flash Player are:

1. No dependency on the browsers. So it's just Adobe using available application development environments for the different platforms, no need to support multiple browsers on each platform, etc. This simplifies the development process of AIR itself, and makes the codebase and its maintenance a lot simpler.
2. Security model is already different. Attack vectors are more traditional as for other natively-running applications: the user needs to install something and then run it, so it's not something you can get compromised with just by visiting a website that happens to embed a dodgy advert.
3. Captive runtime distribution makes compatibility/maintenance less of an issue. There are bugs in the Flash Player that are left in deliberately because to fix them would cause some of the existing web content to break. But in AIR, you can fix things or change things but existing applications would go on working as they tend to be distributed along with the runtime that they were developed against. 

So I think that if Adobe are willing to keep investing in this, it's got a lot longer lifetime. The "if" is an interesting question: I don't know what returns they get on their investments here, so it might be that at some point in the future, they decide it's too unprofitable to continue. At that point, whilst I doubt they will open-source it, I hope that they will make it possible for people to keep using it even if they don't develop the features any further. (In fact, if you have already downloaded an AIR SDK, then you're fine and no longer need Adobe if you wanted to keep developing and distributing AIR-based apps... so this is also a bit of a simpler situation).

**Comment by ROBERT-MCDOWELL** [comment]
just to mention that Adobe works very closely with google and microsoft, today the developers working on CVE issues are essentially from these 2 corporations, so it means that every decision made about the technology is beyond the investment/money fact, but more about control the standards.
btw, I still waiting to see a better technology that flash for the web, html5/javascript just cloned painfully all what flash did since 10 years now.
